### PR TITLE
Accountstore integration

### PR DESCRIPTION
**Pull request by maxme**
First PR to integrate the AccountStore into wpandroid. 

I focused on removing `AccountHelper` from the project, and replace it by the `AccountStore` (injected by dagger). We used  `AccountHelper` a lot of static methods, I had to refactor a bunch of them to remove any call to a stateful component.

Known issues:

* <s>Sign out doesn't work completely (we have to implement `SiteStore.REMOVE_SITES` action)</s> (fixed in ecce4b0)
* <s>Almost all "My Profile" / "Account Settings" features are broken (we have to implement a `AccountStore.POST_SETTINGS` action)</s> (fixed in a4bf98f)
* Only the AccountStore is "wired", so everything else is kind of broken: My sites, <s>Notifications</s> (wired notifications in a60a008)

These are big issues, but this PR is already huge. I'll work on adding the necessary actions in wpstores so we can merge this and then fix those issues.

Note: `feature/wpstores-integration` will be our feature branch and it's been branched from latest `develop`.

### CODE REVIEW

File: WordPress/build.gradle [hunk 1]
```diff
@@ -120,9 +122,18 @@ dependencies {
...
 
+    // Dagger
+    compile 'com.google.dagger:dagger:2.0.2'
+    apt 'com.google.dagger:dagger-compiler:2.0.2'
+    provided 'org.glassfish:javax.annotation:10.0-b28'
+
     // Simperium
     compile 'com.simperium.android:simperium:0.6.8'
 
+    compile ('com.github.wordpress-mobile:WordPress-Stores-Android:tmp-test-max-SNAPSHOT') {
```
Comment (maxme) [inline WordPress/build.gradle]: This is a temporary branch, I'm waiting for some PRs to be merged in wpstores before changing that to `develop-SNAPSHOT`

**Comment by maxme** [comment]
cc @tonyr59h @aforcier

**Comment by maxme** [comment]
Updated with the latest changes from `wpstores` (stores-annotations)

**Comment by aforcier** [comment]
Here's what I've tested so far and my findings:

✅  Sign in to WP.com with incorrect username
✅  Sign in to WP.com with correct login (worked once, but not again)
✅  Change public display name in My Profile
✅  Change public display name remotely, then open 'My Profile' again to check that it updated
✅  Create a new site as a logged-in user (the site is created, but of course not shown in the list since My Sites isn't implemented yet)
❌  Sign in to self-hosted site (though I specified the path to the `xmlrpc.php` file) - the login screen just shows the loading animation continuously. Log:

```
org.wordpress.android.beta D/WordPress-API: Dispatching action: SiteAction-FETCH_SITES_XML_RPC
org.wordpress.android.beta D/WordPress-API: Dispatching action: SiteAction-UPDATE_SITES
org.wordpress.android.beta I/WordPress-NUX: org.wordpress.android.stores.store.SiteStore$OnSiteChanged@963f477
```

I didn't dig in too much since we have to hook up the self-hosted discovery tools anyhow, and we'll find out what's going on then.

**Comment by aforcier** [comment]
Upon successful sign-in, I'm seeing this in the log:

```
android.app.IntentReceiverLeaked: Activity org.wordpress.android.ui.accounts.SignInActivity has leaked IntentReceiver com.simperium.android.WebSocketManager$1@e1fc2c that was originally registered here. Are you missing a call to unregisterReceiver()?
	at android.app.LoadedApk$ReceiverDispatcher.<init>(LoadedApk.java:918)
	at android.app.LoadedApk.getReceiverDispatcher(LoadedApk.java:719)
	at android.app.ContextImpl.registerReceiverInternal(ContextImpl.java:1172)
	at android.app.ContextImpl.registerReceiver(ContextImpl.java:1152)
	at android.app.ContextImpl.registerReceiver(ContextImpl.java:1146)
	at android.content.ContextWrapper.registerReceiver(ContextWrapper.java:554)
	at com.simperium.android.WebSocketManager.<init>(WebSocketManager.java:98)
	at com.simperium.android.AndroidClient.buildChannelProvider(AndroidClient.java:100)
	at com.simperium.android.AndroidClient.buildChannelProvider(AndroidClient.java:28)
	at com.simperium.Simperium.<init>(Simperium.java:65)
	at com.simperium.Simperium.newClient(Simperium.java:33)
	at com.simperium.Simperium.newClient(Simperium.java:29)
	at org.wordpress.android.ui.notifications.utils.SimperiumUtils.configureSimperium(SimperiumUtils.java:46)
	at org.wordpress.android.ui.accounts.SignInFragment.onAuthenticationChanged(SignInFragment.java:923)
	at java.lang.reflect.Method.invoke(Native Method)
	at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:485)
	at org.greenrobot.eventbus.EventBus.postToSubscription(EventBus.java:416)
	at org.greenrobot.eventbus.EventBus.postSingleEventForEventType(EventBus.java:397)
	at org.greenrobot.eventbus.EventBus.postSingleEvent(EventBus.java:370)
	at org.greenrobot.eventbus.EventBus.post(EventBus.java:251)
	at org.wordpress.android.stores.Dispatcher.emitChange(Dispatcher.java:42)
	at org.wordpress.android.stores.store.Store.emitChange(Store.java:18)
	at org.wordpress.android.stores.store.AccountStore$1.onResponse(AccountStore.java:355)
	at org.wordpress.android.stores.store.AccountStore$1.onResponse(AccountStore.java:348)
	at org.wordpress.android.stores.network.rest.wpcom.auth.Authenticator$TokenRequest.deliverResponse(Authenticator.java:107)
	at org.wordpress.android.stores.network.rest.wpcom.auth.Authenticator$PasswordRequest.deliverResponse(Authenticator.java:124)
	at org.wordpress.android.stores.network.rest.wpcom.auth.Authenticator$TokenRequest.deliverResponse(Authenticator.java:89)
	at com.android.volley.ExecutorDelivery$ResponseDeliveryRunnable.run(ExecutorDelivery.java:99)
	at android.os.Handler.handleCallback(Handler.java:739)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:148)
	at android.app.ActivityThread.main(ActivityThread.java:5422)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)

```

**Comment by aforcier** [comment]
Had one crash when attempting to create a new user, the user was not created:

```
FATAL EXCEPTION: main
  Process: org.wordpress.android.beta, PID: 2272
  java.lang.NullPointerException: Attempt to read from field 'byte[] com.android.volley.NetworkResponse.data' on a null object reference
	  at org.wordpress.android.stores.network.rest.wpcom.account.AccountRestClient$8.onErrorResponse(AccountRestClient.java:184)
	  at com.android.volley.Request.deliverError(Request.java:590)
	  at org.wordpress.android.stores.network.rest.wpcom.WPComGsonRequest.deliverError(WPComGsonRequest.java:49)
	  at com.android.volley.ExecutorDelivery$ResponseDeliveryRunnable.run(ExecutorDelivery.java:101)
	  at android.os.Handler.handleCallback(Handler.java:739)
	  at android.os.Handler.dispatchMessage(Handler.java:95)
	  at android.os.Looper.loop(Looper.java:148)
	  at android.app.ActivityThread.main(ActivityThread.java:5422)
	  at java.lang.reflect.Method.invoke(Native Method)
	  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
	  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

Subsequent attempts with a clean install worked, I couldn't reproduce this again.

**Comment by aforcier** [comment]
Fixed a tiny typo in a TODO so it's not missed in a future search: https://github.com/wordpress-mobile/WordPress-Android/pull/4334/commits/7d3f823d5f0ebdacfb3243e5b1153dcbd59273f5

**Comment by aforcier** [comment]
I tested 'upgrading' from current `develop` (logged into a WP.com account) to this branch. On opening the app, I get this error and am shown the login screen:

```
org.wordpress.android E/WordPress-API: NETWORK_XMLRPC - Incorrect username or password. [Code: 403]
	org.xmlrpc.android.XMLRPCFault: Incorrect username or password. [Code: 403]
		at org.xmlrpc.android.XMLRPCClient.parseXMLRPCResponse(XMLRPCClient.java:334)
		at org.xmlrpc.android.XMLRPCClient$Caller.callXMLRPC(XMLRPCClient.java:515)
		at org.xmlrpc.android.XMLRPCClient$Caller.access$000(XMLRPCClient.java:430)
		at org.xmlrpc.android.XMLRPCClient.call(XMLRPCClient.java:231)
		at org.xmlrpc.android.XMLRPCClient.call(XMLRPCClient.java:215)
		at org.xmlrpc.android.ApiHelper$RefreshBlogContentTask.doInBackground(ApiHelper.java:290)
		at org.xmlrpc.android.ApiHelper$RefreshBlogContentTask.doInBackground(ApiHelper.java:232)
		at android.os.AsyncTask$2.call(AsyncTask.java:295)
		at java.util.concurrent.FutureTask.run(FutureTask.java:237)
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
		at java.lang.Thread.run(Thread.java:818)
```

I also seem to end up with two login screens, when I login once there's another login screen underneath and I have to press sign in again to finally enter the app. When I do login though, my sites are still there and everything seems to work normally.

**Comment by maxme** [comment]
> I tested 'upgrading' from current develop (logged into a WP.com account) to this branch

Yes, I skipped all db migration for now, will do when we have finished the first integration steps.

We could eventually migrate the auth token only first.

**Comment by maxme** [comment]
> Sign in to a self-hosted site (though I specified the path to the xmlrpc.php file) - the login screen just shows the loading animation continuously. Log:

This is now fixed in 09912bc

**Comment by maxme** [comment]
Also started [this checklist](https://github.com/wordpress-mobile/WordPress-Android/issues/4394) for integration related tasks.

**Comment by maxme** [comment]
> Had one crash when attempting to create a new user, the user was not created:

Opened a new issue [here](https://github.com/wordpress-mobile/WordPress-Stores-Android/issues/65).

**Comment by maxme** [comment]
> Upon successful sign-in, I'm seeing this in the log:

**Comment by maxme** [comment]
Summary:
- [x] Sign in to a self-hosted site (though I specified the path to the xmlrpc.php file) - the login screen just shows the loading animation continuously. (Fixed in 5e32eff and 09912bc)
- [x] Had one crash when attempting to create a new user. (Fixed in https://github.com/wordpress-mobile/WordPress-Stores-Android/pull/66)
- [x] I tested 'upgrading' from current develop (logged into a WP.com account) to this branch. (**not fixed**, postponed)
- [x] Upon successful sign-in, I'm seeing this in the log `SignInActivity has leaked IntentReceiver com.simperium.android.WebSocketManager blahblahblah` (fixed in ec9f4dd)

**Comment by aforcier** [comment]
All issues I mentioned are fixed now, or on the integration checklist for future PRs: https://github.com/wordpress-mobile/WordPress-Android/issues/4394

:shipit:

**Pull request merged**